Building quality apps/Unit tests.md

```diff
@@ -0,0 +1,155 @@
+
+* Using Mockito for mocking
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+verify(authTokenManger).saveAuthToken("Token")
+```
+With mocks we are verifying a specific behavior which can be seen in the above example using the verify method from Mockito.
+
+* Using Mockito for stubbing
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+whenever(authTokenManager.getAuthToken()).thenReturn("Token")
+```
+With stubs we predefine results for a specific method so that just like it is shown in the above example. Every time the `getAuthToken` method is called we will get a stubbed result which is in our case `"Token"`.
+
+
+It is important to know this terminology because it is essentially part of the  testing fundamentals and it will definitely help you on your journey in becoming better in writing tests. You will also often find these terms in documentation of some testing library that you will use.
+
+### Write testable code
+
+This one is easier said than done. It is a good practice to actually think about your tests when writing production code. In general it is easy to get a picture how you will test something. If you ever get stuck and you are not sure how to test something that you actually wrote, then it is a right moment to consult a colleague or research the problem in more detail. 
+
+Some general guideline when writing testable code is to:
+
+* **Use dependancy injection** - you will hear this a lot but one can not stress enough how important this is. This approach allows us to write loosely coupled code which, in turn, enables us to replace production dependencies with appropriate test doubles. As an end result this allows us to isolate units of work and write good unit tests.
+* **Use abstraction** - it is a good rule of thumb to abstract an implementation in case you are not sure if you can easily test it or not. For example, the use of abstraction allows us to hide a concrete class that uses android specific implementation. Since android is not available on JVM it makes sense to hide it behind an abstraction which we can easily mock in our tests. 
+* **Avoid static code** - static classes, methods and values should not be your first choice when developing something because they are hard to mock and they tend to create global state. In general when dealing with global state it is hard, but not impossible, to create reliable tests. Sometimes you can not avoid using static code which is totally fine but you have to know the implications that this code leaves on your tests. More on that in the next section.
+ 
+
+### Tests should not depend on order execution
+
+As you might already know, in jUnit each test class can have more than just one test and each time a test is executed the test class is initialised. This behaviour is on purpose and it is something that you want to retain because you do not want an older state to mess with tests that are yet to be executed.
```

Comment:
```suggestion
As you might already know, in jUnit each test class can have more than just one test and each time a test is executed the test class is initialised. This behaviour is on purpose and it is something that you want to retain because you do not want an older state to mess with tests that are yet to be executed. **Remember, unit tests are isolated!**
```